File: backend-api-deprecated/src/utils/queryOptimizer.ts
import { Prisma } from '@prisma/client'
import logger from '../config/logger'

/**
 * 查询优化工具类
 * 提供常见的查询优化模式和最佳实践
 */
export class QueryOptimizer {
  /**
   * 构建分页查询参数（偏移分页）
   * 适用于小到中等数据量的分页查询
   * 
   * @param page 页码（从 1 开始）
   * @param pageSize 每页大小
   * @returns Prisma 分页参数
   */
  static buildOffsetPagination(page: number, pageSize: number) {
    const skip = (page - 1) * pageSize
    return {
      skip,
      take: pageSize
    }
  }

  /**
   * 构建游标分页查询参数
   * 适用于大数据量的分页查询，性能更好
   * 
   * @param cursor 游标（上一页最后一条记录的 ID）
   * @param pageSize 每页大小
   * @returns Prisma 游标分页参数
   */
  static buildCursorPagination(cursor: string | undefined, pageSize: number) {
    return {
      take: pageSize + 1, // 多取一条用于判断是否有下一页
      cursor: cursor ? { id: cursor } : undefined,
      skip: cursor ? 1 : 0 // 跳过游标本身
    }
  }

  /**
   * 处理游标分页结果
   * 
   * @param items 查询结果
   * @param pageSize 每页大小
   * @returns 分页结果和下一页游标
   */
  static processCursorPaginationResult<T extends { id: string }>(
    items: T[],
    pageSize: number
  ): {
    items: T[]
    hasMore: boolean
    nextCursor: string | null
  } {
    const hasMore = items.length > pageSize
    const resultItems = hasMore ? items.slice(0, pageSize) : items
    const nextCursor = hasMore && resultItems.length > 0 
      ? resultItems[resultItems.length - 1].id 
      : null

    return {
      items: resultItems,
      hasMore,
      nextCursor
    }
  }

  /**
   * 构建字段选择参数
   * 只查询需要的字段，减少数据传输量
   * 
   * @param fields 字段名数组
   * @returns Prisma select 参数
   */
  static buildFieldSelection(fields: string[]): Record<string, boolean> | undefined {
    if (!fields || fields.length === 0) {
      return undefined
    }

    return fields.reduce((acc, field) => {
      acc[field] = true
      return acc
    }, {} as Record<string, boolean>)
  }

  /**
   * 构建日期范围查询条件
   * 
   * @param field 日期字段名
   * @param startDate 开始日期
   * @param endDate 结束日期
   * @returns Prisma where 条件
   */
  static buildDateRangeFilter(
    field: string,
    startDate?: Date,
    endDate?: Date
  ): Record<string, any> {
    const filter: Record<string, any> = {}

    if (startDate || endDate) {
      filter[field] = {}
      if (startDate) {
        filter[field].gte = startDate
      }
      if (endDate) {
        filter[field].lte = endDate
      }
    }

    return filter
  }

  /**
   * 构建全文搜索查询条件（PostgreSQL）
   * 使用 to_tsvector 和 to_tsquery 进行全文搜索
   * 
   * @param searchTerm 搜索词
   * @param fields 搜索字段数组
   * @returns Prisma raw SQL 查询
   */
  static buildFullTextSearch(searchTerm: string, fields: string[]): Prisma.Sql {
    if (!searchTerm || fields.length === 0) {
      return Prisma.sql`TRUE`
    }

    // 清理搜索词，防止 SQL 注入
    const cleanTerm = searchTerm.replace(/[^\w\s]/g, ' ').trim()
    if (!cleanTerm) {
      return Prisma.sql`TRUE`
    }

    // 构建 COALESCE 字段连接
    const fieldConcat = fields
      .map(field => `COALESCE("${field}", '')`)
      .join(` || ' ' || `)

    return Prisma.sql`to_tsvector('simple', ${Prisma.raw(fieldConcat)}) @@ to_tsquery('simple', ${cleanTerm + ':*'})`
  }

  /**
   * 构建批量查询的 IN 条件
   * 自动分批处理大量 ID，避免 SQL 参数过多
   * 
   * @param ids ID 数组
   * @param batchSize 每批大小，默认 1000
   * @returns 分批后的 ID 数组
   */
  static splitIntoBatches<T>(items: T[], batchSize: number = 1000): T[][] {
    const batches: T[][] = []
    for (let i = 0; i < items.length; i += batchSize) {
      batches.push(items.slice(i, i + batchSize))
    }
    return batches
  }

  /**
   * 记录慢查询
   * 
   * @param queryName 查询名称
   * @param duration 执行时间（毫秒）
   * @param params 查询参数
   */
  static logSlowQuery(queryName: string, duration: number, params?: any): void {
    if (duration > 1000) {
      logger.warn('Slow query detected', {
        queryName,
        duration: `${duration}ms`,
        params
      })
    }
  }

  /**
   * 执行带性能监控的查询
   * 
   * @param queryName 查询名称
   * @param queryFn 查询函数
   * @returns 查询结果
   */
  static async executeWithMonitoring<T>(
    queryName: string,
    queryFn: () => Promise<T>
  ): Promise<T> {
    const startTime = Date.now()
    try {
      const result = await queryFn()
      const duration = Date.now() - startTime
      this.logSlowQuery(queryName, duration)
      return result
    } catch (error) {
      const duration = Date.now() - startTime
      logger.error('Query execution error', {
        queryName,
        duration: `${duration}ms`,
        error
      })
      throw error
    }
  }

  /**
   * 构建排序参数
   * 
   * @param sortBy 排序字段
   * @param sortOrder 排序方向
   * @returns Prisma orderBy 参数
   */
  static buildOrderBy(
    sortBy?: string,
    sortOrder: 'asc' | 'desc' = 'desc'
  ): Record<string, 'asc' | 'desc'> | undefined {
    if (!sortBy) {
      return undefined
    }

    return {
      [sortBy]: sortOrder
    }
  }

  /**
   * 优化 COUNT 查询
   * 对于大表，使用估算值而不是精确计数
   * 
   * @param tableName 表名
   * @param useEstimate 是否使用估算值
   * @returns COUNT 查询 SQL
   */
  static buildCountQuery(tableName: string, useEstimate: boolean = false): Prisma.Sql {
    if (useEstimate) {
      // 使用 PostgreSQL 统计信息估算行数（快速但不精确）
      return Prisma.sql`
        SELECT reltuples::bigint AS estimate
        FROM pg_class
        WHERE relname = ${tableName}
      `
    } else {
      // 精确计数（慢但准确）
      return Prisma.sql`SELECT COUNT(*) FROM ${Prisma.raw(`"${tableName}"`)}`
    }
  }

  /**
   * 构建批量更新的事务
   * 
   * @param updates 更新操作数组
   * @returns Prisma 事务数组
   */
  static buildBatchUpdateTransaction<T>(
    updates: Array<{ where: any; data: any }>,
    model: any
  ): any[] {
    return updates.map(({ where, data }) => 
      model.update({ where, data })
    )
  }

  /**
   * 优化关联查询
   * 使用 include 而不是多次单独查询
   * 
   * @param includes 关联字段数组
   * @returns Prisma include 参数
   */
  static buildInclude(includes: string[]): Record<string, boolean> | undefined {
    if (!includes || includes.length === 0) {
      return undefined
    }

    return includes.reduce((acc, field) => {
      acc[field] = true
      return acc
    }, {} as Record<string, boolean>)
  }

  /**
   * 构建复合索引友好的查询条件
   * 确保查询条件的顺序与索引列顺序一致
   * 
   * @param conditions 查询条件对象
   * @param indexOrder 索引列顺序
   * @returns 重新排序的查询条件
   */
  static optimizeForCompositeIndex(
    conditions: Record<string, any>,
    indexOrder: string[]
  ): Record<string, any> {
    const optimized: Record<string, any> = {}
    
    // 按索引顺序添加条件
    for (const field of indexOrder) {
      if (conditions[field] !== undefined) {
        optimized[field] = conditions[field]
      }
    }
    
    // 添加其他条件
    for (const [key, value] of Object.entries(conditions)) {
      if (!indexOrder.includes(key)) {
        optimized[key] = value
      }
    }
    
    return optimized
  }
}

/**
 * 查询性能分析器
 * 用于分析和优化查询性能
 */
export class QueryAnalyzer {
  private static queryStats: Map<string, {
    count: number
    totalDuration: number
    maxDuration: number
    minDuration: number
  }> = new Map()

  /**
   * 记录查询统计
   * 
   * @param queryName 查询名称
   * @param duration 执行时间
   */
  static recordQuery(queryName: string, duration: number): void {
    const stats = this.queryStats.get(queryName) || {
      count: 0,
      totalDuration: 0,
      maxDuration: 0,
      minDuration: Infinity
    }

    stats.count++
    stats.totalDuration += duration
    stats.maxDuration = Math.max(stats.maxDuration, duration)
    stats.minDuration = Math.min(stats.minDuration, duration)

    this.queryStats.set(queryName, stats)
  }

  /**
   * 获取查询统计报告
   * 
   * @returns 统计报告
   */
  static getReport(): Array<{
    queryName: string
    count: number
    avgDuration: number
    maxDuration: number
    minDuration: number
  }> {
    const report: Array<any> = []

    for (const [queryName, stats] of this.queryStats.entries()) {
      report.push({
        queryName,
        count: stats.count,
        avgDuration: stats.totalDuration / stats.count,
        maxDuration: stats.maxDuration,
        minDuration: stats.minDuration === Infinity ? 0 : stats.minDuration
      })
    }

    // 按平均执行时间降序排序
    return report.sort((a, b) => b.avgDuration - a.avgDuration)
  }

  /**
   * 重置统计数据
   */
  static reset(): void {
    this.queryStats.clear()
  }

  /**
   * 获取慢查询列表
   * 
   * @param threshold 慢查询阈值（毫秒）
   * @returns 慢查询列表
   */
  static getSlowQueries(threshold: number = 1000): Array<any> {
    const report = this.getReport()
    return report.filter(stat => stat.avgDuration > threshold)
  }
}

export default QueryOptimizer

File: backend-api/app/utils/query_optimizer.py
"""
查询优化工具模块

提供常见的查询优化模式和最佳实践，包括：
- 分页查询优化
- 批量操作优化
- 预加载关联数据
- 慢查询监控
- 查询性能分析

与 Node.js 后端的 queryOptimizer.ts 保持功能一致。
"""
from typing import Any, Dict, List, Optional, TypeVar, Generic, Callable
from sqlalchemy import select, func, text
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload, joinedload
from datetime import datetime
import logging
import time

logger = logging.getLogger(__name__)

T = TypeVar('T')


class PaginationResult(Generic[T]):
    """分页查询结果"""
    
    def __init__(
        self,
        items: List[T],
        total: int,
        page: int,
        page_size: int
    ):
        self.items = items
        self.total = total
        self.page = page
        self.page_size = page_size
        self.total_pages = (total + page_size - 1) // page_size if page_size > 0 else 0
        self.has_next = page < self.total_pages
        self.has_prev = page > 1
    
    def to_dict(self) -> Dict[str, Any]:
        """转换为字典格式"""
        return {
            "items": self.items,
            "pagination": {
                "total": self.total,
                "page": self.page,
                "pageSize": self.page_size,
                "totalPages": self.total_pages,
                "hasNext": self.has_next,
                "hasPrev": self.has_prev
            }
        }


class CursorPaginationResult(Generic[T]):
    """游标分页查询结果"""
    
    def __init__(
        self,
        items: List[T],
        has_more: bool,
        next_cursor: Optional[str] = None
    ):
        self.items = items
        self.has_more = has_more
        self.next_cursor = next_cursor
    
    def to_dict(self) -> Dict[str, Any]:
        """转换为字典格式"""
        return {
            "items": self.items,
            "hasMore": self.has_more,
            "nextCursor": self.next_cursor
        }


class QueryOptimizer:
    """
    查询优化工具类
    提供常见的查询优化模式和最佳实践
    """
    
    @staticmethod
    def build_offset_pagination(page: int, page_size: int) -> Dict[str, int]:
        """
        构建偏移分页参数
        适用于小到中等数据量的分页查询
        
        Args:
            page: 页码（从 1 开始）
            page_size: 每页大小
            
        Returns:
            包含 offset 和 limit 的字典
        """
        offset = (page - 1) * page_size
        return {
            "offset": offset,
            "limit": page_size
        }
    
    @staticmethod
    async def paginate_query(
        db: AsyncSession,
        query: Any,
        page: int,
        page_size: int
    ) -> PaginationResult:
        """
        执行分页查询
        
        Args:
            db: 数据库会话
            query: SQLAlchemy 查询对象
            page: 页码
            page_size: 每页大小
            
        Returns:
            分页结果
        """
        # 获取总数
        count_query = select(func.count()).select_from(query.subquery())
        total_result = await db.execute(count_query)
        total = total_result.scalar() or 0
        
        # 获取分页数据
        pagination = QueryOptimizer.build_offset_pagination(page, page_size)
        paginated_query = query.offset(pagination["offset"]).limit(pagination["limit"])
        result = await db.execute(paginated_query)
        items = result.scalars().all()
        
        return PaginationResult(
            items=items,
            total=total,
            page=page,
            page_size=page_size
        )
    
    @staticmethod
    def build_cursor_pagination(
        cursor: Optional[str],
        page_size: int
    ) -> Dict[str, Any]:
        """
        构建游标分页参数
        适用于大数据量的分页查询，性能更好
        
        Args:
            cursor: 游标（上一页最后一条记录的 ID）
            page_size: 每页大小
            
        Returns:
            游标分页参数
        """
        return {
            "cursor": cursor,
            "limit": page_size + 1,  # 多取一条用于判断是否有下一页
            "skip": 1 if cursor else 0  # 跳过游标本身
        }
    
    @staticmethod
    def process_cursor_pagination_result(
        items: List[Any],
        page_size: int,
        id_field: str = "id"
    ) -> CursorPaginationResult:
        """
        处理游标分页结果
        
        Args:
            items: 查询结果
            page_size: 每页大小
            id_field: ID 字段名
            
        Returns:
            游标分页结果
        """
        has_more = len(items) > page_size
        result_items = items[:page_size] if has_more else items
        next_cursor = getattr(result_items[-1], id_field) if has_more and result_items else None
        
        return CursorPaginationResult(
            items=result_items,
            has_more=has_more,
            next_cursor=next_cursor
        )
    
    @staticmethod
    def build_date_range_filter(
        field: str,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None
    ) -> Dict[str, Any]:
        """
        构建日期范围查询条件
        
        Args:
            field: 日期字段名
            start_date: 开始日期
            end_date: 结束日期
            
        Returns:
            查询条件字典
        """
        filters = {}
        if start_date:
            filters[f"{field}_gte"] = start_date
        if end_date:
            filters[f"{field}_lte"] = end_date
        return filters
    
    @staticmethod
    def split_into_batches(items: List[T], batch_size: int = 1000) -> List[List[T]]:
        """
        将列表分批处理
        自动分批处理大量数据，避免 SQL 参数过多
        
        Args:
            items: 数据列表
            batch_size: 每批大小，默认 1000
            
        Returns:
            分批后的列表数组
        """
        batches = []
        for i in range(0, len(items), batch_size):
            batches.append(items[i:i + batch_size])
        return batches
    
    @staticmethod
    def log_slow_query(
        query_name: str,
        duration: float,
        params: Optional[Dict[str, Any]] = None
    ) -> None:
        """
        记录慢查询
        慢查询阈值：1000ms（与 Node.js 后端一致）
        
        Args:
            query_name: 查询名称
            duration: 执行时间（毫秒）
            params: 查询参数
        """
        if duration > 1000:
            logger.warning(
                f"Slow query detected: {query_name}",
                extra={
                    "queryName": query_name,
                    "duration": f"{duration:.2f}ms",
                    "params": params
                }
            )
    
    @staticmethod
    async def execute_with_monitoring(
        query_name: str,
        query_fn: Callable,
        *args,
        **kwargs
    ) -> Any:
        """
        执行带性能监控的查询
        
        Args:
            query_name: 查询名称
            query_fn: 查询函数
            *args: 位置参数
            **kwargs: 关键字参数
            
        Returns:
            查询结果
        """
        start_time = time.time()
        try:
            result = await query_fn(*args, **kwargs)
            duration = (time.time() - start_time) * 1000  # 转换为毫秒
            QueryOptimizer.log_slow_query(query_name, duration)
            return result
        except Exception as e:
            duration = (time.time() - start_time) * 1000
            logger.error(
                f"Query execution error: {query_name}",
                extra={
                    "queryName": query_name,
                    "duration": f"{duration:.2f}ms",
                    "error": str(e)
                }
            )
            raise
    
    @staticmethod
    def build_preload_options(relationships: List[str], strategy: str = "selectinload"):
        """
        构建预加载选项
        使用预加载避免 N+1 查询问题
        
        Args:
            relationships: 关系字段列表
            strategy: 加载策略，"selectinload" 或 "joinedload"
            
        Returns:
            预加载选项列表
        """
        load_strategy = selectinload if strategy == "selectinload" else joinedload
        return [load_strategy(getattr(model, rel)) for rel in relationships]
    
    @staticmethod
    async def batch_create(
        db: AsyncSession,
        model_class: Any,
        data_list: List[Dict[str, Any]]
    ) -> List[Any]:
        """
        批量创建记录
        优化数据库写入性能
        
        Args:
            db: 数据库会话
            model_class: 模型类
            data_list: 数据字典列表
            
        Returns:
            创建的对象列表
        """
        objects = [model_class(**data) for data in data_list]
        db.add_all(objects)
        await db.flush()
        return objects
    
    @staticmethod
    async def batch_update(
        db: AsyncSession,
        model_class: Any,
        updates: List[Dict[str, Any]],
        id_field: str = "id"
    ) -> int:
        """
        批量更新记录
        
        Args:
            db: 数据库会话
            model_class: 模型类
            updates: 更新数据列表，每项包含 id 和要更新的字段
            id_field: ID 字段名
            
        Returns:
            更新的记录数
        """
        count = 0
        for update_data in updates:
            record_id = update_data.pop(id_field)
            stmt = (
                model_class.__table__.update()
                .where(getattr(model_class, id_field) == record_id)
                .values(**update_data)
            )
            result = await db.execute(stmt)
            count += result.rowcount
        return count
    
    @staticmethod
    def build_order_by(
        sort_by: Optional[str] = None,
        sort_order: str = "desc"
    ) -> Optional[str]:
        """
        构建排序参数
        
        Args:
            sort_by: 排序字段
            sort_order: 排序方向，"asc" 或 "desc"
            
        Returns:
            排序参数
        """
        if not sort_by:
            return None
        return f"{sort_by} {sort_order}"
    
    @staticmethod
    async def execute_in_batches(
        db: AsyncSession,
        items: List[Any],
        process_fn: Callable,
        batch_size: int = 1000
    ) -> List[Any]:
        """
        分批执行操作
        
        Args:
            db: 数据库会话
            items: 数据列表
            process_fn: 处理函数
            batch_size: 批次大小
            
        Returns:
            处理结果列表
        """
        results = []
        batches = QueryOptimizer.split_into_batches(items, batch_size)
        
        for batch in batches:
            batch_result = await process_fn(db, batch)
            results.extend(batch_result)
        
        return results


class QueryAnalyzer:
    """
    查询性能分析器
    用于分析和优化查询性能
    """
    
    _query_stats: Dict[str, Dict[str, Any]] = {}
    
    @classmethod
    def record_query(cls, query_name: str, duration: float) -> None:
        """
        记录查询统计
        
        Args:
            query_name: 查询名称
            duration: 执行时间（毫秒）
        """
        if query_name not in cls._query_stats:
            cls._query_stats[query_name] = {
                "count": 0,
                "total_duration": 0.0,
                "max_duration": 0.0,
                "min_duration": float('inf')
            }
        
        stats = cls._query_stats[query_name]
        stats["count"] += 1
        stats["total_duration"] += duration
        stats["max_duration"] = max(stats["max_duration"], duration)
        stats["min_duration"] = min(stats["min_duration"], duration)
    
    @classmethod
    def get_report(cls) -> List[Dict[str, Any]]:
        """
        获取查询统计报告
        
        Returns:
            统计报告列表
        """
        report = []
        
        for query_name, stats in cls._query_stats.items():
            avg_duration = stats["total_duration"] / stats["count"] if stats["count"] > 0 else 0
            report.append({
                "queryName": query_name,
                "count": stats["count"],
                "avgDuration": round(avg_duration, 2),
                "maxDuration": round(stats["max_duration"], 2),
                "minDuration": round(stats["min_duration"], 2) if stats["min_duration"] != float('inf') else 0
            })
        
        # 按平均执行时间降序排序
        report.sort(key=lambda x: x["avgDuration"], reverse=True)
        return report
    
    @classmethod
    def reset(cls) -> None:
        """重置统计数据"""
        cls._query_stats.clear()
    
    @classmethod
    def get_slow_queries(cls, threshold: float = 1000.0) -> List[Dict[str, Any]]:
        """
        获取慢查询列表
        
        Args:
            threshold: 慢查询阈值（毫秒），默认 1000ms
            
        Returns:
            慢查询列表
        """
        report = cls.get_report()
        return [stat for stat in report if stat["avgDuration"] > threshold]


# 导出主要类和函数
__all__ = [
    "QueryOptimizer",
    "QueryAnalyzer",
    "PaginationResult",
    "CursorPaginationResult"
]
